exatomic/adf/tape21.py

```diff
@@ -0,0 +1,362 @@
+from exatomic.core.tensor import JCoupling, NMRShielding
+from exatomic.base import z2sym, sym2isomass
+import numpy as np
+import pandas as pd
+import six
+import warnings
+
+class Tape21Meta(TypedMeta):
+    atom = Atom
+    frequency = Frequency
+    gradient = Gradient
+    j_coupling = JCoupling
+    nmr_shielding = NMRShielding
+
+class Tape21(six.with_metaclass(Tape21Meta, Editor)):
+    '''
+    Parser for ADF Tape21 that have been converted to an ASCII file with
+    their dmpkf utility.
+
+    **All properties are parsed based on the input order.**
+
+    Note:
+        This is not yet tested for ADF versions newer than 2017.
+    '''
+
+    @staticmethod
+    def rmass_mwc(data, symbol):
+        '''
+        Calculate the reduced masses from the mass-weighted normal modes. With
+        the equation,
+
+        .. math::
+            \\mu_i = \\left(\\sum_k^{3N} \\left(\\frac{l_{MWCk,i}}
+                                                {\\sqrt{m_k}}\\right)^2\\right)^{-1}
+
+        Note:
+            This assumes that the normal modes have already been placed in the
+            :code:`['dx', 'dy', 'dz']` columns.
+
+        Args:
+            data (:class:`pandas.DataFrame`): Data frame the has the mass-weighted
+                                              normal modes.
+            symbol (:obj:`list`): List-like object that has the atomic symbols.
+
+        Returns:
+            r_mass (:class:`numpy.ndarray`): Array containing the calculated reduced
+                                             masses.
+        '''
+        cols = ['dx', 'dy', 'dz']
+        mapper = sym2isomass(symbol)
+        mass = list(map(mapper.get, symbol))
+        mass = np.repeat(mass, 3).astype(float)
+        mass = mass.reshape(data[cols].shape)
+        disps = data[cols].values
+        r_mass = np.sum(np.square(disps)/mass)
+        r_mass = 1/r_mass
+        return r_mass
+
+    @staticmethod
+    def rmass_cart(data, symbol):
+        '''
+        Calculate the reduced masses from the normalized non-mass-weighted cartesian
+        normal modes. With the equation,
+
+        .. math::
+            \\mu_i = \\left(\\sum_k^{3N} l_{CARTk,i}^2\\right)^{-1}
+
+        Note:
+            This assumes that the normal modes have already been placed in the
+            :code:`['dx', 'dy', 'dz']` columns.
+
+        Args:
+            data (:class:`pandas.DataFrame`): Data frame the has the non-mass-weighted
+                                              cartesian normal modes.
+            symbol (:obj:`list`): List-like object that has the atomic symbols.
+
+        Returns:
+            r_mass (:class:`numpy.ndarray`): Array containing the calculated reduced
+                                             masses.
+        '''
+        cols = ['dx', 'dy', 'dz']
+        # get the isotopic masses of the unique atoms
+        mapper = sym2isomass(symbol)
+        # get a list of the isotopic masses
+        mass = list(map(mapper.get, symbol))
```

Comment:
The pattern of `mapper =...; mass = ...` and related lines repeat themselves. I would consider pulling out this functionality into a separate function and converting these static methods into generalized functions elsewhere in the codebase ( in core perhaps?). Not a blocker for me.